pkg/controller/machineset.go

```diff
@@ -323,22 +324,39 @@ func (c *controller) manageReplicas(allMachines []*v1alpha1.Machine, machineSet
 		return nil
 	}
 
-	var activeMachines, staleMachines []*v1alpha1.Machine
+	var activeMachines, runningMachines, staleMachines []*v1alpha1.Machine
 	for _, machine := range allMachines {
 		if IsMachineActive(machine) {
-			//klog.Info("Active machine: ", machine.Name)
 			activeMachines = append(activeMachines, machine)
+
+			if IsMachineRunning(machine) {
+				runningMachines = append(runningMachines, machine)
+			}
 		} else if IsMachineFailed(machine) {
 			staleMachines = append(staleMachines, machine)
 		}
 	}
 
+	machineDeletionWindow := c.calculateStaleMachineDeletionWindow(activeMachines, runningMachines, staleMachines, machineSet)
+
 	if len(staleMachines) >= 1 {
 		klog.V(2).Infof("Deleting stale machines")
 	}
-	c.terminateMachines(staleMachines, machineSet)
+	deletedStaleMachines, _ := c.terminateStaleMachines(staleMachines, machineDeletionWindow, machineSet)
+	staleMachinesLeft := len(staleMachines) - deletedStaleMachines
 
 	diff := len(activeMachines) - int(machineSet.Spec.Replicas)
+	if staleMachinesLeft > 0 {
+		// Count the leftover stale machines against .spec.replicas to prevent sudden surge in machines which can happen if large number of stale machines are left over
+		diff -= staleMachinesLeft
```

Comment:
If `diff` is already negative, the further substraction will increase the number. This step should either be below in the if block or some other alternative.
Simple example from Gaurav: 
```
active machines = 12, spec = 15, deleted 2, left over = 1,
we should now be creating only 2 machines
….and not 4
```
@zuzzas Can you please take care of this part?
credits @ggaurav10

Reply:
Taken care of.